components/tools/OmeroWeb/test/integration/test_api_wells.py

```diff
@@ -0,0 +1,231 @@
+
+
+def assert_objects(conn, json_objects, omero_ids_objects, dtype="Project",
+                   group='-1', extra=None, opts=None):
+    """
+    Load objects from OMERO, via conn.getObjects().
+
+    marshal with omero_marshal and compare with json_objects.
+    omero_ids_objects can be IDs or list of omero.model objects.
+
+    @param: extra       List of dicts containing expected extra json data
+                        e.g. {'omero:childCount': 1}
+    """
+    pids = []
+    for p in omero_ids_objects:
+        try:
+            pids.append(long(p))
+        except TypeError:
+            pids.append(p.id.val)
+    conn.SERVICE_OPTS.setOmeroGroup(group)
+    objs = conn.getObjects(dtype, pids, respect_order=True, opts=opts)
+    objs = [p._obj for p in objs]
+    expected = marshal_objects(objs)
+    assert len(json_objects) == len(expected)
+    for i, o1, o2 in zip(range(len(expected)), json_objects, expected):
+        if extra is not None and i < len(extra):
+            o2.update(extra[i])
+        # dumping to json and loading (same as test data) means that
+        # unicode has been handled in same way, e.g. Pixel size symbols.
+        o2 = json.loads(json.dumps(o2))
+        # remove any urls from json (tested elsewhere)
+        remove_urls(o1)
+
+        assert o1 == o2
+
+
+class TestWells(IWebTest):
+    """Tests querying Wells."""
+
+    @pytest.fixture()
+    def user1(self):
+        """Return a new user in a read-annotate group."""
+        group = self.new_group(perms='rwra--')
+        return self.new_client_and_user(group=group)
+
+    def create_plate_wells(self, user1, rows, cols):
+        """Return Plate with Wells."""
+        updateService = get_update_service(user1)
+        plate = PlateI()
+        plate.name = rstring('plate')
+        plate = updateService.saveAndReturnObject(plate)
+
+        # Create Wells for plate
+        for row in range(rows):
+            for col in range(cols):
+                # create Well
+                well = WellI()
+                well.column = rint(col)
+                well.row = rint(row)
+                well.plate = PlateI(plate.id.val, False)
+                # Only wells in first Column have well-samples etc
+                if col == 0:
+                    # Have 3 images/well-samples in these wells
+                    for i in range(3):
+                        image = self.create_test_image(
+                            size_x=5, size_y=5, session=user1[0].getSession())
+                        ws = WellSampleI()
+                        ws.image = ImageI(image.id, False)
+                        ws.well = well
+                        ws.posX = LengthI(i * 10, UnitsLength.REFERENCEFRAME)
+                        ws.posY = LengthI(i, UnitsLength.REFERENCEFRAME)
+                        well.addWellSample(ws)
+                updateService.saveObject(well)
+        return plate
+
+    @pytest.fixture()
+    def small_plate(self, user1):
+        """
```

Comment:
Done in https://github.com/openmicroscopy/openmicroscopy/pull/5047/commits/84ad329202a8d9294dcef9ef3e7a7eff21399f38